### PR TITLE
 Warning summary respects -nowarn

### LINKED ISSUE #11952: Scala 2.13.2 bridge builds with warnings
## reproduction steps
Any fresh build of a project with 2.13.2.

## problem
I get
```
[info] Non-compiled module 'compiler-bridge_2.13' for Scala 2.13.2. Compiling...
9 warnings
```
It's only occurring at the first bridge build

## expectation
No warnings displayed

### PR DESCRIPTION
**Pull request by som-snytt**
Previously, hasWarning meant a warning had been displayed
(as opposed to counted). Since last refactor, it means
counted, so warnings were summarized under -nowarn.
Until 2.13.1, there was no summary under -nowarn.

Let -nowarn mean -Xmaxwarns 0 (to mean show me no warnings)
and additionally don't summarize any warnings (including
erroring under -Werror).

Fixes scala/bug#11952

### CODE REVIEW

**Comment by lrytz** [comment]
Some tests failed because summaries no longer show in repl tests

**Comment by som-snytt** [comment]
Edit: fixed just the REPL output, which was due to `withSuppressedWarnings` also tamping `maxwarns` but not restoring it.

**Comment by som-snytt** [comment]
It still says 1 warning, which is better than 9.
```
[info] Non-compiled module 'compiler-bridge_2.13' for Scala 2.13.3-20200429-100920-dd427d7. Compiling...
1 warning
```

**Comment by som-snytt** [comment]
@lrytz That leaking warning is because `reportSuspendedMessages` (which sets `suppressionsComplete`) is at end of typer phase, so a warning from patmat (if not "suppressed") is issued immediately. Why does that happen after typer and not after the run, at `summarize` time?

The division of state and responsibility between `PerRunReporting` and the `Reporter` is confusing to me.

Probably just make `ConsoleReporter` not issue summary under `-nowarn`, then the warning count is irrelevant.

**Comment by som-snytt** [comment]
Also picks up a commit to check that compiler options to a partest `DirectTest` are kosher.

Noticed while linting `run` tests; maybe there is a `neg` test that intentionally tests bad args?

**Comment by lrytz** [comment]
> That leaking warning is because `reportSuspendedMessages` (which sets `suppressionsComplete`) is at end of typer phase, so a warning from patmat (if not "suppressed") is issued immediately.

I don't really follow what happens here. `suppressionsComplete` is about `@nowarn` annotations within source code, which we only collect during typer. So to avoid displaying warnings that would be hidden by a `@nowarn`, all warnings are buffered until after typer.

All that should not have anything to do with `settins.nowarn`. Are you saying that some patmat warning is getting issued in 2.13.2 under `-nowarn`?

> The division of state and responsibility between `PerRunReporting` and the `Reporter` is confusing to me.

I think the division exists because of being within or outside the global cake. But I see the point: it would probably be good to move everything to `Reporter` that doesn't need access to `Symbol`, and make `RunReporting` a few convenience methods.

File: src/compiler/scala/tools/nsc/reporters/ConsoleReporter.scala [hunk 1]
```diff
@@ -25,9 +25,11 @@ class ConsoleReporter(val settings: Settings, val reader: BufferedReader, val wr
   def doReport(pos: Position, msg: String, severity: Severity): Unit = display(pos, msg, severity)
 
   override def finish(): Unit = {
-    import reflect.internal.util.StringOps.countElementsAsString
-    if (hasWarnings) echo(countElementsAsString(warningCount, WARNING.toString.toLowerCase))
-    if (hasErrors)   echo(countElementsAsString(errorCount, ERROR.toString.toLowerCase))
+    if (!settings.nowarn) {
+      import reflect.internal.util.StringOps.countElementsAsString
+      if (hasWarnings) echo(countElementsAsString(warningCount, WARNING.toString.toLowerCase))
+      if (hasErrors)   echo(countElementsAsString(errorCount, ERROR.toString.toLowerCase))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/reporters/ConsoleReporter.scala]: the error summary should probably stay even with `-nowarn`?

**Comment by som-snytt** [comment]
Yes, patmat issues warning, debug stack trace:
```
        at scala.reflect.internal.Reporter.increment(Reporting.scala:121)
        at scala.reflect.internal.Reporter.filteredInfo(Reporting.scala:115)
        at scala.reflect.internal.Reporter.warning(Reporting.scala:110)
        at scala.tools.nsc.Reporting$PerRunReporting.issueWarning(Reporting.scala:116)
        at scala.tools.nsc.Reporting$PerRunReporting.checkSuppressedAndIssue(Reporting.scala:130)
        at scala.tools.nsc.Reporting$PerRunReporting.warning(Reporting.scala:251)
        at scala.tools.nsc.Reporting$PerRunReporting.warning(Reporting.scala:255)
        at scala.tools.nsc.typechecker.Contexts$ContextReporter.warning(Contexts.scala:1638)
        at scala.tools.nsc.typechecker.Contexts$Context.warning(Contexts.scala:813)
        at scala.tools.nsc.transform.patmat.Interface$MatchMonadInterface.reportMissingCases(PatternMatching.scala:196)
```
It is "counted", so `ConsoleReporter` on `finish` summarizes it.

This PR adds two checks for `-nowarn` at `PerRunReporter.summarize` and also `ConsoleReporter`. This doesn't seem terrible to me, as the semantics are simply, "Don't print warning-related messages, including summaries."

But if `-nowarn` sets `-Xmaxwarns 0` why are there any warnings? Because `maxwarns` only tells `reporters.Reporter` to cap the warnings that are displayed; the count is "upstream" in `internal.Reporter`.

As a footnote, not only cake per se but the split due to `reflect` is ongoing confusion. By confusion, I mean I couldn't understand the whole picture as a casual contributor.

File: src/compiler/scala/tools/nsc/reporters/ConsoleReporter.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/reporters/ConsoleReporter.scala]: You mean error reporting? Yes, I had a second thought about that. I'll update to retain it. That was an example of hitting my time budget and patience budget.

**Comment by som-snytt** [comment]
`[error] Caused by: sbt.ForkMain$ForkError: java.io.IOException: No space left on device`

**Review by lrytz — APPROVED** [review]
Thank you for giving so much of your patience and time budgeds!

**Pull request merged**